### PR TITLE
#1900 - Add type parameters for zonotope

### LINKED ISSUE #1900: Add type parameters for zonotope center and generators
See https://github.com/JuliaReach/LazySets.jl/issues/481.

### PR DESCRIPTION
**Pull request by mforets**
Closes #1900.

### CODE REVIEW

File: src/Sets/Zonotope.jl [hunk 1]
```diff
@@ -29,25 +29,40 @@ segments.
...
-Below we illustrate both ways.
+Zonotopes can be constructed in two different ways: either passing the generators
+as a matrix, where each column represents a generator, or passing a list of vectors
+where each vector represents a generator. Below we illustrate both ways.
 
 The optional argument `remove_zero_generators` controls whether we remove zero
 columns from the `generators` matrix. This option is active by default.
 
+
+
```
Comment (schillic) [inline src/Sets/Zonotope.jl]:
```suggestion
```

File: src/Sets/Zonotope.jl [hunk 2]
```diff
@@ -75,40 +90,54 @@ vectors, each vector representing a generator:
...
+function Zonotope(center::VN, generators_list::AbstractVector{VN};
+                  remove_zero_generators::Bool=true) where {N<:Real, VN<:AbstractVector{N}}
+    n = length(center)
+    p = length(generators_list)
+    G = Matrix{N}(undef, n, p) # TODO: generalize type, see #2040
+    if remove_zero_generators
+        for (i, gi) in enumerate(generators_list)
+            iszero(gi) && continue
+            @inbounds G[:, i] = gi
+        end
```
Comment (schillic) [inline src/Sets/Zonotope.jl]: This does not work for zero generators (there will be `undef` columns). I would just not change the function in this PR, it is unrelated. In any case, you should add a test.

Comment (mforets) [reply src/Sets/Zonotope.jl]: okay... i didn't understand what is your use case that doesn't work, but ill defer to a follow up issue :thumbsup:

Comment (schillic) [reply src/Sets/Zonotope.jl]: You create a matrix with `p` columns but only insert at column `i` if the vector is not zero. Try calling this function with a single zero vector and you will get an `undef` matrix.

File: src/convert.jl [hunk 3]
```diff
@@ -701,7 +701,7 @@ The cartesian product is obtained by:
   in the off-diagonal; for this reason, the generator matrix of the returned
   zonotope is built as a sparse matrix.
 """
-function convert(::Type{Zonotope}, cp::CartesianProduct{N, Zonotope{N}, Zonotope{N}}) where {N<:Real}
+function convert(::Type{Zonotope}, cp::CartesianProduct{N, <:Zonotope{N}, <:Zonotope{N}}) where {N<:Real}
```
Comment (schillic) [inline src/convert.jl]: The docstring and the entry in the docs need to be updated.

**Comment by mforets** [comment]
> So we should check whether Reachability still works with this changeset.

i've opened https://github.com/JuliaReach/Reachability.jl/issues/747. fwiw, i don't plan to work on this any time soon though; although i'm fine to review... i'm writing the adapted algorithms in [ReachabilityAnalysis.jl](https://github.com/mforets/ReachabilityAnalysis.jl)

File: docs/src/man/reach_zonotopes_hybrid.md [hunk 4]
```diff
@@ -40,7 +40,8 @@ using Plots, LazySets, LinearAlgebra
 
 function reach_hybrid(As, Ts, init, δ, μ, T, max_order, instant_transitions)
     # initialize queue with initial mode and states at time t=0
-    queue = [(init[1], init[2], 0.)]
+    queue = Vector{Tuple{Zonotope, Integer, Float64}}(undef, 1)
+    queue[1] = [(init[1], init[2], 0.0)]
```
Comment (mforets) [inline docs/src/man/reach_zonotopes_hybrid.md]:
```suggestion
    queue[1] = (init[1], init[2], 0.0)
```

File: docs/src/lib/conversion.md [hunk 5]
```diff
@@ -37,7 +37,7 @@ convert(::Type{Zonotope}, ::AbstractHyperrectangle)
 convert(::Type{Zonotope}, ::AbstractZonotope)
 convert(::Type{IntervalArithmetic.IntervalBox}, ::AbstractHyperrectangle)
 convert(::Type{Hyperrectangle}, ::IntervalArithmetic.IntervalBox)
-convert(::Type{Zonotope}, ::CartesianProduct{N, Zonotope{N}, Zonotope{N}}) where {N<:Real}
+convert(::Type{Zonotope}, ::CartesianProduct{N, Z1, Z2}) where {N<:Real, Z1<:Zonotope{N}, Z2<:Zonotope{N}}
```
Comment (mforets) [inline docs/src/lib/conversion.md]:
```suggestion
convert(::Type{Zonotope}, ::CartesianProduct{N, <:Zonotope{N}, <:Zonotope{N}}) where {N<:Real}
```

File: docs/src/lib/conversion.md [hunk 6]
```diff
@@ -37,7 +37,7 @@ convert(::Type{Zonotope}, ::AbstractHyperrectangle)
 convert(::Type{Zonotope}, ::AbstractZonotope)
 convert(::Type{IntervalArithmetic.IntervalBox}, ::AbstractHyperrectangle)
 convert(::Type{Hyperrectangle}, ::IntervalArithmetic.IntervalBox)
-convert(::Type{Zonotope}, ::CartesianProduct{N, Zonotope{N}, Zonotope{N}}) where {N<:Real}
+convert(::Type{Zonotope}, ::CartesianProduct{N, ZN1, ZN2}) where {N<:Real, ZN1<:Zonotope{N}, ZN2<:Zonotope{N}})
```
Comment (mforets) [inline docs/src/lib/conversion.md]:
```suggestion
convert(::Type{Zonotope}, ::CartesianProduct{N, ZN1, ZN2}) where {N<:Real, ZN1<:Zonotope{N}, ZN2<:Zonotope{N}}
```

**Pull request merged**